File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/insert_interval.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval() : start(-1), end(-1) {}

  Interval(int start, int end) : start(start), end(end) {}
};

// O(n) / O(1)
vector<Interval> insert_interval(vector<Interval> &a, Interval &new_interval) {
  vector<Interval> res;

  int i = 0;
  while (i < a.size() && new_interval.start > a[i].end) {
    res.push_back(a[i]);
    i++;
  }

  while (i < a.size() && a[i].start <= new_interval.end) {
    new_interval.start = min(new_interval.start, a[i].start);
    new_interval.end = max(new_interval.end, a[i].end);
    i++;
  }

  res.push_back(new_interval);

  while (i < a.size()) {
    res.push_back(a[i]);
    i++;
  }

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, start, end;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a(n);
    for (int i = 0; i < n; i++) {
      cin >> start >> end;
      a[i] = Interval(start, end);
    }

    cin >> start >> end;
    Interval new_interval(start, end);

    auto res = insert_interval(a, new_interval);
    for (auto r : res) {
      cout << r.start << " " << r.end << "\n";
    }
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/InsertInterval.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class InsertInterval {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Interval> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
          int start = in.nextInt();
          int end = in.nextInt();
          a.add(new Interval(start, end));
        }

        int start = in.nextInt();
        int end = in.nextInt();
        Interval newInterval = new Interval(start, end);

        List<Interval> res = insertInterval(a, newInterval);
        for (var r : res) {
          System.out.printf("%d %d\n", r.start, r.end);
        }
      }
    }
  }

  // O(n) / O(1)
  public static List<Interval> insertInterval(List<Interval> a,
                                              Interval newInterval) {
    List<Interval> res = new ArrayList<>();

    int i = 0;
    while (i < a.size() && a.get(i).end < newInterval.start) {
      res.add(a.get(i));
      i++;
    }

    while (i < a.size() && a.get(i).start <= newInterval.end) {
      newInterval.start = Math.min(newInterval.start, a.get(i).start);
      newInterval.end = Math.max(newInterval.end, a.get(i).end);
      i++;
    }

    res.add(newInterval);

    while (i < a.size()) {
      res.add(a.get(i));
      i++;
    }

    return res;
  }
}